chord_metadata_service/restapi/middleware.py

```diff
@@ -1,86 +1,104 @@
+        self.authorize_datasets = 'd100,d200'
+        """
+        
+        if settings.CANDIG_AUTHORIZATION == 'OPA' and request.method == 'GET'\
+            and any(re.match(path_re, request.path) for path_re in self.authorized_paths):
+            if settings.CACHE_TIME != 0:
+                error_response = {
+                    "error": "cache time needs to be zero to be secure"
+                }
+                response = HttpResponseServerError(json.dumps(error_response))
+                response["Content-Type"] = "application/json"
+                return response
+            token = self.get_opa_token_from_request(request.headers)
+            opa_res_datasets = self.get_opa_res(token, request.path, request.method)
+            if len(opa_res_datasets) == 0:
+                self.authorize_datasets = 'NO_DATASETS_AUTHORIZED'
+            elif type(opa_res_datasets) == tuple and opa_res_datasets[0] == "error": # error response
+                return opa_res_datasets[1]
+            else:
+                self.authorize_datasets = ",".join(opa_res_datasets)
+            request.GET = request.GET.copy() # Make request.GET mutable
+            request.GET.update({'authorized_datasets': self.authorize_datasets})
+            response = self.get_response(request)
+            return response
+        response = self.get_response(request)
+        return response
+
+    def get_opa_token_from_request(self, headers):
+        """
+        Extracts token from request's header X-CANDIG-LOCAL-OIDC
+        """
+        token = headers.get("X-CANDIG-LOCAL-OIDC")
+        if token == None:
+            return ""
+        return token.strip('"')
+
+    def get_request_body(self, token, path, method):
+        """
+        Returns request body required to query OPA
+        """
+        return {
+                "input": {
+                    "token": token,
+                    "body": {
+                        "path": path,
+                        "method": method
+                    }
+                }
             }
-        }
 
-        url = f"{settings.CANDIG_OPA_URL}/v1/data/metadata"
+    def get_opa_res(self, token, path, method):
+        """
+        Get allowed dataset result from OPA
+        """
+        try:
+            response = requests.post(settings.CANDIG_OPA_URL +
+                                         "/v1/data/permissions/datasets",
+                                         headers={"Authorization": f"Bearer {settings.CANDIG_OPA_SECRET}"},
+                                         json=self.get_request_body(token, path, method),
+                                         verify=settings.CANDIG_ROOT_CA)
+            response.raise_for_status()
+        except requests.exceptions.RequestException:
+            error_response = {
+                "error": "error getting response from authorization service"
+            }
+            response = HttpResponseServerError(json.dumps(error_response))
```

Comment:
I think it might be better to say, `This request failed because we are unable to retrieve necessary info related to your account. Please contact your system administrator for assistance.`

Instead of a 500, which is `HttpResponseServerError` below, I think it's better to return a 403, which is `HttpResponseForbidden`.